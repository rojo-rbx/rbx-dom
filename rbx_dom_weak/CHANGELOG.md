# rbx_dom_weak Changelog

## Unreleased Changes
* Implemented `Default` for `WeakDom`, useful when using Serde or creating an empty `WeakDom`

## 2.6.0 (2023-10-03)
* Added `WeakDom::clone_multiple_into_external` that allows cloning multiple subtrees all at once into a given `WeakDom`, useful for preserving `Ref` properties that point across cloned subtrees ([#364])

[#364]: https://github.com/rojo-rbx/rbx-dom/pull/364

## 2.5.0 (2023-08-09)
* Fix potential stack overflow when creating or inserting into a `WeakDom`. ([#279])
* Added `InstanceBuilder::has_property` for checking if an `InstanceBuilder` defines a given property. ([#283])
* Added `WeakDom::clone_within` and `WeakDom::clone_into_external` for cloning instances into the same or a different `WeakDom`, respectively. ([#312])

[#279]: https://github.com/rojo-rbx/rbx-dom/pull/279
[#283]: https://github.com/rojo-rbx/rbx-dom/pull/283
[#312]: https://github.com/rojo-rbx/rbx-dom/pull/312

## 2.4.0 (2022-06-05)
* Added `WeakDom::into_raw` for enabling fast, non-tree-preserving transformations.
* Added `empty`, `with_class`, and `set_class` methods to `InstanceBuilder`.

## 2.3.0 (2021-10-11)
* Updated to rbx_types 1.3.

## 2.2.0 (2021-07-19)
* Updated to rbx_types 1.2.

## 2.1.0 (2021-07-02)
* Updated to rbx_types 1.1.

## 2.0.0 (2021-06-26) (yanked)
* Updated to rbx_types 1.0.

## 2.0.0-alpha.1 (2021-02-16)
This release is a major, breaking change that introduces many fixes and features.

* `RbxTree` was replaced with `WeakDom`
	* Changed `new` to accept an `InstanceBuilder`, which can contain children.
	* Renamed `get_root_id` to `root_ref`.
	* Renamed `get_instance` to `get_by_ref`.
	* Renamed `get_instance_mut` to `get_by_ref_mut`.
	* Renamed `move_instance` to `transfer`.
	* Renamed `set_parent` to `transfer_within`.
	* Renamed `insert_instance` to `insert`. This method now accepts an `InstanceBuilder`.
	* Removed `remove_instance`, replaced by `destroy` and `transfer`.
	* Removed `iter_all_ids`.
	* Removed `descendants`.
	* Added `root` and `root_mut` for accessing the root instance directly.
* `RbxInstanceProperties` was replaced with `InstanceBuilder`
	* This API is completely different, as it's now a builder. It is now much easier to construct instances.
* `RbxInstance` was replaced with `Instance`
	* This type now directly exposes its values instead of implementing `Deref` for another type.
* Moved types into the `rbx_types` crate, re-exported as `rbx_dom_weak::types`.
	* `RbxId` was replaced with `types::Ref`, which can now represent null referents.
* Added `DomViewer` API from rbx_dom_test to make testing instance trees easier.

## 1.10.1 (2019-12-18)
* Updated `base64`, `md5`, and `uuid` dependencies

## 1.10.0 (2019-09-15)
* Added `RbxTree::set_parent`, for moving instances within a single tree.

## 1.9.0 (2019-07-12)
* Changed `BrickColor` to be much more correct
	* All BrickColor values are now correctly available instead of just palette colors
	* Enum values now align with the `BrickColor.Number` property and `BrickColor.new` constructor
	* `BrickColor` no longer serializes to strings in human-readable Serde formats to avoid information loss. These colors have colliding names:
		* Rust
		* Lilac
		* Gold
		* Deep orange

## 1.8.2 (2019-07-10)
* Added conversion from `Int32` to `BrickColor`. This should fix serialization of `SpawnLocation` instances, which use `int` instead of `BrickColor` in at least the XML model format.

## 1.8.1 (2019-06-20)
* Fixed Serde being able to deserialize `RbxValue` and `RbxValueType`'s internal `__Nonexhaustive` value. This should also fix Serde recommending this value to users in error messages.
* Fixed `BrickColor` serialization to use Roblox color names

## 1.8.0 (2019-06-10)
* Added the `SharedString` type, which is used by Roblox to reduce redundant copies of binary buffers. ([#63](https://github.com/rojo-rbx/rbx-dom/pull/63))

## 1.7.0 (2019-05-14)
* Added conversions from `BrickColor` to `Color3` and `Color3uint8`. ([#46](https://github.com/rojo-rbx/rbx-dom/pull/46))
* Added conversions from `Color3` to `Color3uint8` and vice-versa.
* Changed the human readable serialization of `BinaryString` values to be base64-encoded. This makes JSON-encoded values much smaller. This changes the details of the unspecified serialization of rbx_dom_weak.
* `RbxValueConversion` now derives `PartialEq`.

## 1.6.0 (2019-05-12)
* Added `RbxTree::move_instance` API to move instances from one tree to another.
* Fixed `RbxTree::descendants` to no longer return the instance of the ID you give it. This may break code relying on this (broken) assumption, but was definitely a bug.
* `RbxValueConversion` now derives `Debug` and `Clone` ([#52](https://github.com/rojo-rbx/rbx-dom/issues/52))

## 1.5.0 (2019-05-06)
* Added support for `BrickColor` ([#29](https://github.com/rojo-rbx/rbx-dom/pull/29))
* Added `RbxValue::try_convert_ref`, which can be used to try to coerce values. ([#42](https://github.com/rojo-rbx/rbx-dom/pull/42))
* `Content` values can now be inferred from string literals.
* `RbxValueType` now derives Serde's `Serialize` and `Deserialize` traits

## 1.4.0 (2019-03-27)
* Added support for new types:
	* `NumberRange`
	* [#15](https://github.com/rojo-rbx/rbx-dom/pull/15):
		* `Int64`
		* `Float64`
	* [#20](https://github.com/rojo-rbx/rbx-dom/pull/20):
		* `Rect`
		* `Ray`
		* `ColorSequence`
		* `NumberSequence`

## 1.3.0 (2019-03-14)
* Added `sort_children_by_key` and `sort_children_unstable_by_key` methods to `RbxInstance` to reorder children safely

## 1.2.0 (2019-03-13)
* `RbxTree` and `RbxInstance` are now clonable without an `unimplemented` panic
	* Cloned trees preserve their IDs as-is and there's no public API yet to transplant instances between trees. This is mostly useful for comparing trees before/after a mutation, which we're using in Rojo.

## 1.1.0 (2019-03-11)
* Marked `PhysicalProperties` as `pub`
* Fixed `Serialize` impl for `UnresolvedRbxValue`
	* This removes `Serialize` from `AmbiguousRbxValue`, which should not break any real code.

## 1.0.0 (2019-03-01)
* `RbxValue` and `RbxValueType` can no longer be matched exhaustively, which enables adding new types without breaking code in the future
* Added support for `UDim`, `UDim2`, and `Content` value types
* `PhysicalProperties` is no longer a stub type

## 0.3.1 (2019-02-26)
* Added support for bare bool values when deserializing `UnresolvedRbxValue`
* Implemented `Serialize` for `UnresolvedRbxValue` and `AmbiguousRbxValue`
* Implemented `From<RbxValue>` for `UnresolvedRbxValue`

## 0.3.0 (2019-02-14)
* Renamed crate from `rbx_tree` to `rbx_dom_weak`
* Added support for `Ref` values ([#8](https://github.com/rojo-rbx/rbx-dom/pull/8))
* Added `UnresolvedRbxValue` and `AmbiguousRbxValue`, intended to be used alongside `rbx_reflection` to make specifying values less verbose.

## 0.2.0 (2019-01-25)
* Added new variants for `RbxValue`:
	* Int32
	* Float32
	* Enum
	* Vector2
	* Color3
	* Color3uint8
	* Vector3int16
	* Vector2int16
	* CFrame
	* PhysicalProperties (Stub)

## 0.1.0
* Initial release
* Supports `String`, `Bool`, and `Vector3` property values
