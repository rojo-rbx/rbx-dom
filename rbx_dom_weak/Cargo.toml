[package]
name = "rbx_dom_weak"
description = "Weakly-typed Roblox DOM implementation for Rust"
version = "2.4.0"
license = "MIT"
documentation = "https://docs.rs/rbx_dom_weak"
homepage = "https://github.com/rojo-rbx/rbx-dom"
repository = "https://github.com/rojo-rbx/rbx-dom.git"
readme = "README.md"
authors = ["Lucien Greathouse <me@lpghatguy.com>"]
edition = "2018"

[dependencies]
rbx_types = { version = "1.4.0", path = "../rbx_types", features = ["serde"] }

serde = "1.0.137"

[dev-dependencies]
insta = { version="1.14.1", features=["yaml"] }
