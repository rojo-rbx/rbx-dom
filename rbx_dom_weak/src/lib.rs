//! rbx_dom_weak is a common representation of the Roblox DOM for Rust. It's
//! designed to play nicely with the borrow checker and allows accessing
//! instances by ID in constant time.
//!
//! Constructing a new tree of instances is accomplished by first creating an
//! [`InstanceBuilder`] object that describes a tree of instances and then
//! wrapping it with an [`WeakDom`]:
//!
//! ```
//! use rbx_dom_weak::{InstanceBuilder, WeakDom};
//!
//! let dm = InstanceBuilder::new("DataModel");
//!
//! let mut dom = WeakDom::new(dm);
//!
//! println!("ID of DOM root is {:?}", dom.root_ref());
//! ```
//!
//! Once we have a tree, we can use [`WeakDom::insert`] and
//! [`WeakDom::get_by_ref`] to add instances to the tree and retrieve them.
//!
//! ```
//! use rbx_dom_weak::{InstanceBuilder, WeakDom};
//!
//! let mut dom = WeakDom::new(InstanceBuilder::new("DataModel"));
//!
//! // We can define properties using any type that can be converted to an
//! // rbx_dom_weak::types::Variant.
//! let http_service = InstanceBuilder::new("HttpService")
//!     .with_property("HttpEnabled", true);
//!
//! let http_service_id = dom.insert(dom.root_ref(), http_service);
//!
//! println!("HttpService has ID {:?}", http_service_id);
//! ```
//!
//! To change properties on an instance that's already present in the tree, use
//! [`WeakDom::get_by_ref_mut`]. Note that it isn't possible to add or remove
//! children through this method, use [`WeakDom::insert`] and
//! [`WeakDom::destroy`] instead.

#![deny(missing_docs)]

mod dom;
mod instance;
mod viewer;

pub use rbx_types as types;

pub use ahash::AHashMap;

pub use crate::{
    dom::WeakDom,
    instance::{Instance, InstanceBuilder},
    viewer::{DomViewer, ViewedInstance},
};

/// Helper trait that provides convenience methods for `AHashMap`.
pub trait HashMapExt {
    /// Constructs an empty map.
    fn new() -> Self;

    /// Constructs an empty map with at least the specified capacity.
    fn with_capacity(capacity: usize) -> Self;
}

impl<K, V> HashMapExt for AHashMap<K, V> {
    /// Creates an empty `AHashMap` using the default value for its hasher.
    fn new() -> Self {
        AHashMap::default()
    }

    /// Creates an empty `AHashMap` with at least the specified capacity using
    /// the default value for its hasher.
    fn with_capacity(capacity: usize) -> Self {
        AHashMap::with_capacity_and_hasher(capacity, Default::default())
    }
}
