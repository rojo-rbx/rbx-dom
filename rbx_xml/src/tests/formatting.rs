//! Test to ensure the formatting of files does not change.

use rbx_dom_weak::DomViewer;

const INPUT: &str = r#"<roblox version="4">
    <Item class="TestClass" referent="Parent">
        <Properties>
            <Axes name="TestAxes"><axes>5</axes></Axes>
            <BinaryString name="TestBinaryString">SGVsbG8sIHdvcmxkIQ==</BinaryString>
            <bool name="TestBool">true</bool>
            <CoordinateFrame name="TestCFrame">
                <X>123</X>
                <Y>456</Y>
                <Z>789</Z>
                <R00>987</R00>
                <R01>654</R01>
                <R02>432</R02>
                <R10>210</R10>
                <R11>0</R11>
                <R12>-12345</R12>
                <R20>765</R20>
                <R21>234</R21>
                <R22>123123</R22>
            </CoordinateFrame>
            <ColorSequence name="TestColorSequence">0 0 0.5 1 0 1 1 0.5 0 0 </ColorSequence>
            <Color3 name="TestColor3">
                <R>1</R>
                <G>0.5</G>
                <B>125600</B>
            </Color3>
            <Content name="TestContent1"><url>Wow!</url></Content>
            <Content name="TestContent2"><null></null></Content>
        </Properties>
        <Item class="TestClass" referent="Child1">
            <Properties>
                <double name="TestDouble">INF</double>
                <Faces name="TestFace"><faces>42</faces></Faces>
                <float name="TestFloat">NAN</float>
                <Font name="TestFont">
                    <Family><url>Font Family</url></Family>
                    <Weight>100</Weight>
                    <Style>Normal</Style>
                    <CachedFaceId><null></null></CachedFaceId>
                </Font>
                <int name="TestInt">1337</int>
                <int64 name="TestInt64">8675309</int64>
            </Properties>
        </Item>
        <Item class="TestClass" referent="Child2">
            <Properties>
                <NumberRange name="TestNumberRange">-1337 1337</NumberRange>
                <NumberSequence name="TestNumberSequence">0 10 20 1 30 40</NumberSequence>
                <OptionalCoordinateFrame name="TestOptionCFrame1"></OptionalCoordinateFrame>
                <OptionalCoordinateFrame name="TestOptionCFrame2">
                    <CFrame>
                        <X>100</X>
                        <Y>200</Y>
                        <Z>300</Z>
                        <R00>-100</R00>
                        <R01>-200</R01>
                        <R02>-300</R02>
                        <R10>123</R10>
                        <R11>456</R11>
                        <R12>-123</R12>
                        <R20>-456</R20>
                        <R21>INF</R21>
                        <R22>-INF</R22>
                    </CFrame>
                </OptionalCoordinateFrame>
            </Properties>
            <Item class="TestClass" referent="Grandchild">
                <Properties>
                    <PhysicalProperties name="TestPhysicalProperties1">
                        <CustomPhysics>false</CustomPhysics>
                    </PhysicalProperties>
                    <PhysicalProperties name="TestPhysicalProperties2">
                        <CustomPhysics>true</CustomPhysics>
                        <Density>1</Density>
                        <Friction>-1</Friction>
                        <Elasticity>0.15625</Elasticity>
                        <FrictionWeight>-0.15625</FrictionWeight>
                        <ElasticityWeight>NAN</ElasticityWeight>
                    </PhysicalProperties>
                    <ProtectedString name="TestProtectedString">Hello world, again!</ProtectedString>
                    <Ray name="TestRay">
                        <origin>
                            <X>10</X>
                            <Y>20</Y>
                            <Z>30</Z>
                        </origin>
                        <direction>
                            <X>30</X>
                            <Y>20</Y>
                            <Z>10</Z>
                        </direction>
                    </Ray>
                    <Rect2D name="TestRect">
                        <min>
                            <X>1</X>
                            <Y>2</Y>
                        </min>
                        <max>
                            <X>0.0</X>
                            <Y>INF</Y>
                        </max>
                    </Rect2D>
                    <Ref name="TestRef1">Parent</Ref>
                    <Ref name="TestRef2">null</Ref>
                    <SharedString name="TestSharedString">TestHash</SharedString>
                    <string name="TestString">Hello, world!</string>
                    <token name="TestEnum">1337</token>
                    <UDim name="TestUDim">
                        <S>1234.5</S>
                        <O>-123</O>
                    </UDim>
                    <UDim2 name="TestUDim2">
                        <XS>1234.5</XS>
                        <XO>-123</XO>
                        <YS>-1234.5</YS>
                        <YO>123</YO>
                    </UDim2>
                    <UniqueId name="TestUniqueId1">00000000000000000000000000000000</UniqueId>
                    <UniqueId name="TestUniqueId2">1234567890abcdef00c0ffeebadf00d0</UniqueId>
                    <Vector2 name="TestVector2">
                        <X>INF</X>
                        <Y>0</Y>
                    </Vector2>
                    <Vector3 name="TestVector3">
                        <X>0</X>
                        <Y>INF</Y>
                        <Z>123</Z>
                    </Vector3>
                    <Vector3int16 name="TestVector3int16">
                        <X>-10</X>
                        <Y>0</Y>
                        <Z>10</Z>
                    </Vector3int16>
                </Properties>
            </Item>
        </Item>
    </Item>
    <SharedStrings>
        <SharedString md5="TestHash">SGVsbG8sIHdvcmxkIQ==</SharedString>
    </SharedStrings>
</roblox>"#;

#[test]
fn formatting() {
    let _ = env_logger::try_init();

    let de = crate::from_str(
        INPUT,
        crate::DecodeOptions::new().property_behavior(crate::DecodePropertyBehavior::NoReflection),
    )
    .map_err(|e| panic!("cannot deserialize: {}", e))
    .unwrap();

    insta::assert_yaml_snapshot!("deserialized", DomViewer::new().view_children(&de));

    let mut ser = Vec::with_capacity(INPUT.len());
    crate::to_writer(
        &mut ser,
        &de,
        de.root().children(),
        crate::EncodeOptions::new().property_behavior(crate::EncodePropertyBehavior::NoReflection),
    )
    .map_err(|e| panic!("cannot serialize: {}", e))
    .unwrap();

    let ser_str = String::from_utf8(ser)
        .map_err(|e| panic!("serialized result not string: {}", e))
        .unwrap();

    insta::assert_snapshot!("serialized", ser_str)
}
