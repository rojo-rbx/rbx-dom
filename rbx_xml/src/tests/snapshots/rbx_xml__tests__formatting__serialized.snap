---
source: rbx_xml/src/tests/formatting.rs
expression: ser_str
---
<roblox version="4">
  <Item class="TestClass" referent="0">
    <Properties>
      <string name="Name">TestClass</string>
      <Axes name="TestAxes">
        <axes>5</axes>
      </Axes>
      <BinaryString name="TestBinaryString"><![CDATA[SGVsbG8sIHdvcmxkIQ==]]></BinaryString>
      <bool name="TestBool">true</bool>
      <CoordinateFrame name="TestCFrame">
        <X>123</X>
        <Y>456</Y>
        <Z>789</Z>
        <R00>987</R00>
        <R01>654</R01>
        <R02>432</R02>
        <R10>210</R10>
        <R11>0</R11>
        <R12>-12345</R12>
        <R20>765</R20>
        <R21>234</R21>
        <R22>123123</R22>
      </CoordinateFrame>
      <Color3 name="TestColor3">
        <R>1</R>
        <G>0.5</G>
        <B>125600</B>
      </Color3>
      <ColorSequence name="TestColorSequence">0 0 0.5 1 0 1 1 0.5 0 0 </ColorSequence>
      <Content name="TestContent1">
        <url>Wow!</url>
      </Content>
      <Content name="TestContent2">
        <null>
        </null>
      </Content>
    </Properties>
    <Item class="TestClass" referent="1">
      <Properties>
        <string name="Name">TestClass</string>
        <double name="TestDouble">INF</double>
        <Faces name="TestFace">
          <faces>42</faces>
        </Faces>
        <float name="TestFloat">NAN</float>
        <Font name="TestFont">
          <Family>
            <url>Font Family</url>
          </Family>
          <Weight>100</Weight>
          <Style>Normal</Style>
          <CachedFaceId>
            <null>
            </null>
          </CachedFaceId>
        </Font>
        <int name="TestInt">1337</int>
        <int64 name="TestInt64">8675309</int64>
      </Properties>
    </Item>
    <Item class="TestClass" referent="2">
      <Properties>
        <string name="Name">TestClass</string>
        <NumberRange name="TestNumberRange">-1337 1337 </NumberRange>
        <NumberSequence name="TestNumberSequence">0 10 20 1 30 40 </NumberSequence>
        <OptionalCoordinateFrame name="TestOptionCFrame1">
        </OptionalCoordinateFrame>
        <OptionalCoordinateFrame name="TestOptionCFrame2">
          <CFrame>
            <X>100</X>
            <Y>200</Y>
            <Z>300</Z>
            <R00>-100</R00>
            <R01>-200</R01>
            <R02>-300</R02>
            <R10>123</R10>
            <R11>456</R11>
            <R12>-123</R12>
            <R20>-456</R20>
            <R21>inf</R21>
            <R22>-inf</R22>
          </CFrame>
        </OptionalCoordinateFrame>
      </Properties>
      <Item class="TestClass" referent="3">
        <Properties>
          <string name="Name">TestClass</string>
          <token name="TestEnum">1337</token>
          <PhysicalProperties name="TestPhysicalProperties1">
            <CustomPhysics>false</CustomPhysics>
          </PhysicalProperties>
          <PhysicalProperties name="TestPhysicalProperties2">
            <CustomPhysics>true</CustomPhysics>
            <Density>1</Density>
            <Friction>-1</Friction>
            <Elasticity>0.15625</Elasticity>
            <FrictionWeight>-0.15625</FrictionWeight>
            <ElasticityWeight>NAN</ElasticityWeight>
          </PhysicalProperties>
          <string name="TestProtectedString">Hello world, again!</string>
          <Ray name="TestRay">
            <origin>
              <X>10</X>
              <Y>20</Y>
              <Z>30</Z>
            </origin>
            <direction>
              <X>30</X>
              <Y>20</Y>
              <Z>10</Z>
            </direction>
          </Ray>
          <Rect2D name="TestRect">
            <min>
              <X>1</X>
              <Y>2</Y>
            </min>
            <max>
              <X>0</X>
              <Y>INF</Y>
            </max>
          </Rect2D>
          <Ref name="TestRef1">0</Ref>
          <Ref name="TestRef2">null</Ref>
          <SharedString name="TestSharedString">7eXAsQ8uxJecabUvYeQv9Q==</SharedString>
          <string name="TestString">Hello, world!</string>
          <UDim name="TestUDim">
            <S>1234.5</S>
            <O>-123</O>
          </UDim>
          <UDim2 name="TestUDim2">
            <XS>1234.5</XS>
            <XO>-123</XO>
            <YS>-1234.5</YS>
            <YO>123</YO>
          </UDim2>
          <UniqueId name="TestUniqueId1">00000000000000000000000000000000</UniqueId>
          <UniqueId name="TestUniqueId2">1234567890abcdef00c0ffeebadf00d0</UniqueId>
          <Vector2 name="TestVector2">
            <X>INF</X>
            <Y>0</Y>
          </Vector2>
          <Vector3 name="TestVector3">
            <X>0</X>
            <Y>INF</Y>
            <Z>123</Z>
          </Vector3>
          <Vector3int16 name="TestVector3int16">
            <X>-10</X>
            <Y>0</Y>
            <Z>10</Z>
          </Vector3int16>
        </Properties>
      </Item>
    </Item>
  </Item>
  <SharedStrings>
    <SharedString md5="7eXAsQ8uxJecabUvYeQv9Q==">SGVsbG8sIHdvcmxkIQ==</SharedString>
  </SharedStrings>
</roblox>
